File: cursor.js
/* One cursor owner for a prepared Xpra connection. Sizes are CSS pixels. */
"use strict";

class FloeXpraCursor {
  constructor(client) {
    this.client = client;
    this.generation = 0;
    this.current = null;
    this.source = null;
    this.pending = null;
    this.disposed = false;
    this.density = this.pixelDensity();
    this.densityChanged = () => {
      if (this.disposed) return;
      const density = this.pixelDensity();
      if (density !== this.density) {
        this.density = density;
        if (this.source) this.render();
      }
      this.watchDensity();
    };
    window.addEventListener("resize", this.densityChanged);
    this.watchDensity();
  }

  pixelDensity() {
    // Integral backing density preserves exact logical dimensions at fractional DPR.
    return Math.min(4, Math.max(1, Math.ceil(window.devicePixelRatio || 1)));
  }

  watchDensity() {
    this.media?.removeEventListener("change", this.densityChanged);
    this.media = window.matchMedia(`(resolution: ${window.devicePixelRatio || 1}dppx)`);
    this.media.addEventListener("change", this.densityChanged);
  }

  cancelPending() {
    if (!this.pending) return;
    this.pending.image.onload = this.pending.image.onerror = null;
    this.pending.image.src = "";
    URL.revokeObjectURL(this.pending.url);
    this.pending = null;
  }

  receive(packet) {
    if (this.disposed) return;
    const generation = ++this.generation;
    this.cancelPending();
    if (packet.length < 9) { this.reset(); return; }
    const [width, height, xhot, yhot] = packet.slice(4, 8);
    const bytes = packet[9];
    if (packet[1] !== "png" || ![width, height, xhot, yhot].every(Number.isInteger) ||
        width < 1 || height < 1 || width > 1024 || height > 1024 ||
        xhot < 0 || yhot < 0 || xhot >= width || yhot >= height ||
        !(bytes instanceof Uint8Array) || !bytes.length || bytes.length > 4 * 1024 * 1024) {
      this.reset(); return;
    }
    const url = URL.createObjectURL(new Blob([bytes], {type:"image/png"}));
    const image = new Image();
    this.pending = {image, url};
    image.onload = () => {
      if (this.disposed || generation !== this.generation) return;
      URL.revokeObjectURL(url);
      this.pending = null;
      image.onload = image.onerror = null;
      if (image.naturalWidth !== width || image.naturalHeight !== height) { this.reset(); return; }
      this.source = {image, width, height, xhot, yhot};
      this.render();
    };
    image.onerror = () => { if (generation === this.generation) this.reset(); };
    image.src = url;
  }

  render() {
    if (this.disposed || !this.source) return;
    const source = this.source;
    const scale = Math.min(1, 24 / Math.max(source.width, source.height));
    const width = Math.max(1, Math.round(source.width * scale));
    const height = Math.max(1, Math.round(source.height * scale));
    const xhot = Math.min(width - 1, Math.round(source.xhot * scale));
    const yhot = Math.min(height - 1, Math.round(source.yhot * scale));
    try {
      const canvas = document.createElement("canvas");
      canvas.width = width * this.density;
      canvas.height = height * this.density;
      const context = canvas.getContext("2d");
      context.imageSmoothingEnabled = true;
      context.imageSmoothingQuality = "high";
      context.drawImage(source.image, 0, 0, canvas.width, canvas.height);
      const url = canvas.toDataURL("image/png");
      this.current = Object.freeze({url, width, height, xhot, yhot,
        css:`image-set(url("${url}") ${this.density}x) ${xhot} ${yhot}, default`});
      this.apply();
    } catch {
      this.reset();
    }
  }

  apply() {
    for (const win of Object.values(this.client.id_to_window)) win.set_cursor(this.current);
    // A shadow pointer uses the same geometry as the CSS cursor. Its next
    // position packet makes it visible again; never leave a stale bitmap showing.
    const shadow = document.querySelector("#shadow_pointer");
    if (shadow) shadow.style.display = "none";
  }

  reset() {
    this.generation++;
    this.cancelPending();
    this.source = null;
    this.current = null;
    this.apply();
  }

  dispose() {
    if (this.disposed) return;
    this.disposed = true;
    window.removeEventListener("resize", this.densityChanged);
    this.media?.removeEventListener("change", this.densityChanged);
    this.reset();
  }
}

File: cursor.go
package nativeapps

import (
	_ "embed"
	"fmt"
	"strings"
)

//go:embed cursor.js
var cursorSource []byte

// All reviewed HTML versions share this cursor owner. The original client still
// owns pointer packets; normalization changes neither coordinates nor buttons.
func prepareCursorHTML(index, client, window []byte) ([]byte, []byte, []byte, error) {
	var failure error
	replace := func(source, old, next string) string {
		if strings.Count(source, old) != 1 {
			failure = fmt.Errorf("unsupported Xpra HTML cursor contract near %.64q", old)
			return source
		}
		return strings.Replace(source, old, next, 1)
	}
	section := func(source, start, end, next string) string {
		i, j := strings.Index(source, start), strings.Index(source, end)
		if i < 0 || j < i || strings.Count(source, start) != 1 || strings.Count(source, end) != 1 {
			failure = fmt.Errorf("unsupported Xpra HTML cursor section near %.64q", start)
			return source
		}
		return source[:i] + next + source[j:]
	}
	h, c, w := string(index), string(client), string(window)
	h = replace(h, "    <script type=\"text/javascript\" src=\"js/Client.js\"></script>", "    <script type=\"text/javascript\" src=\"js/FloeCursor.js\"></script>\n    <script type=\"text/javascript\" src=\"js/Client.js\"></script>")
	c = replace(c, "  _process_hello(packet) {", "  _process_hello(packet) {\n    this.floeCursor?.dispose();\n    this.floeCursor = new FloeXpraCursor(this);")
	c = replace(c, "  close_protocol() {", "  close_protocol() {\n    this.floeCursor?.dispose();\n    this.floeCursor = null;")
	c = replace(c, "    this.id_to_window[wid] = win;", "    this.id_to_window[wid] = win;\n    win.set_cursor(this.floeCursor?.current);")
	c = section(c, "  reset_cursor() {", "  _process_window_icon(packet) {", `  reset_cursor() {
    this.floeCursor?.reset();
  }

  _process_cursor(packet) {
    this.floeCursor?.receive(packet);
  }

`)
	c = section(c, "    let cursor_url;", "    style.display = \"inline\";", `    const cursor = this.floeCursor?.current;
    if (!win || !cursor) { style.display = "none"; return; }
    style.width = `+"`${cursor.width}px`"+`;
    style.height = `+"`${cursor.height}px`"+`;
    shadow_pointer.src = cursor.url;
    style.left = `+"`${x - cursor.xhot}px`"+`;
    style.top = `+"`${y - cursor.yhot}px`"+`;
`)
	w = section(w, "  reset_cursor() {", "  eos() {", `  set_cursor(cursor) {
    this.div.style.cursor = cursor?.css || "default";
  }

`)
	w = replace(w, "    this.cursor_data = null;\n", "")
	if failure != nil {
		return nil, nil, nil, failure
	}
	return []byte(h), []byte(c), []byte(w), nil
}
